adding GECXX to LAPACK

scr2016 opened the pull request:
**This a draft for dgecxx.f. to REVIEW and COMMENT**

scr2016 committed:
Addded a DRAFT of DGECS routine.

scr2016 committed:
Added dgecx.f, a DRAFT for DGECX routine

mgates3 commented on SRC/DGECX.f:
```diff
@@ -0,0 +1,1173 @@
+*>          Specifies how the factors of CX factorization
+*>          are returned.
+*>
+*>          = 'P' or 'p' : return only the column permutaion matrix P
```
Pedantic: LAPACK doesn't normally specify both Uppercase and lowercase. Cf.:
```
lapack/SRC> grep "^\*> += '\w'" lapack/SRC/zpotrf.f
*>          = 'U':  Upper triangle of A is stored;
*>          = 'L':  Lower triangle of A is stored.

lapack/SRC> grep "^\*> += '\w'" lapack/SRC/*.f`
```

mgates3 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*>          Specifies how the factors of CX factorization
+*>          are returned.
+*>
+*>          = 'P' or 'p' : return only the column permutaion matrix P
```
Pedantic: LAPACK doesn't normally specify both Uppercase and lowercase. Cf.:
```
lapack/SRC> grep "^\*> += '\w'" lapack/SRC/zpotrf.f
*>          = 'U':  Upper triangle of A is stored;
*>          = 'L':  Lower triangle of A is stored.

lapack/SRC> grep "^\*> += '\w'" lapack/SRC/*.f`
```

scr2016 replied:
Now, the routine specifies only Uppercase character parameters.

mgates3 reviewed (changes requested):
Some initial comments.

scr2016 committed:
Fix documentation for ABSTOL and factor flags

changed the description of DGECX

scr2016 committed:
Delete SRC/DGECX.f

scr2016 committed:
Delete SRC/dgecx.f

scr2016 committed:
Added  DGECXX routine in dgecxx.f

mgates3 commented on SRC/dgecxx.f:
```diff
@@ -0,0 +1,1437 @@
+*            http://www.netlib.org/lapack/explore-html/
+*
+*> \htmlonly
+*> Download DGEQP3RK + dependencies
```
`GECXX` instead of `GEQP3RK`.

scr2016 replied:
resolved

mgates3 commented on SRC/dgecxx.f:
```diff
@@ -0,0 +1,1437 @@
+*>          On exit, if INFO >= 0, WORK(1) returns the optimal LIWORK.
+*> \endverbatim
+*>
+*> \param[out] LIWORK
```
`\param[in]`

scr2016 replied:
resolved

scr2016 committed:
Edited leading comments in dgecxx.f

scr2016 committed:
Refactor comments and some efficiency improvements

scr2016 committed:
Fix LIWORK calculations and JPIV adjustment comments

scr2016 committed:
Rename subroutine DGECXX22 to DGECXX

scr2016 committed:
Merge branch 'Reference-LAPACK:master' into scr2016-dgecx-draft

scr2016 committed:
dgecxx.f Corrected calculation of check for LDX input parameter and refined calculation for workspace LWORK and LIWORK

scr2016 committed:
dgecxx.f : simplified workspace calculation

scr2016 committed:
dgecxx.f modified the description of LWORK and LIWORK

scr2016 committed:
dgecxx.f: updated description of FACT parameter, and LWORK parameter to include NX crosspver point

scr2016 committed:
dgecxx.f: corrected  the descriptions of QRC, X and LDX parameters

scr2016 committed:
dgecxx.f: corrected the description of LWORK parameter regarding NX crossover point.

scr2016 committed:
dgecxx.f: corrected the LDX minimum value and description from: If FACT = 'X', LDX >= max(1,M) to LDX >= max(1,min(M,N))

scr2016 committed:
dgecxx.f: corrected speclling the comments inside the code

scr2016 committed:
dgecxx.f: version 2 of the factor C generation algorithm

scr2016 committed:
dgecxx.f: commit 68d8fba5c112de973b69445a8faa9d6cd3c4ce3e was incorrect, so I reverted  the description of LDX to now say again  If FACT = 'X', LDX >= max(1,M), plus expanded the description of the parameter X.

scr2016 committed:
dgecxx.f: cahanged the description of X, added notation for lls problem as min|| A - C*X||_F for the unknown X

scr2016 committed:
dgecxx.f: corrected and updated the description of the parameter X.

scr2016 committed:
dgecxx.f: changed the algorithm to generated matrix C, now it is verision 3.

scr2016 committed:
dgecxx.f: corrected comments in the version 3 algorithm  to generate the  matrix C

scr2016 committed:
dgecxx.f: updated the workspace dimension parameter LIWORK description

scr2016 committed:
dgecxx.f: corrected the description of the parameter LIWORL

scr2016 committed:
dgecxx.f: upadated the description of the parameters C and QRC

scr2016 committed:
dgecxx.f: added top comments

scr2016 committed:
dgecxx.f: changed 2 LQUERY and LIQEURY variables into 1 LQUERY

scr2016 committed:
added dgecxx.f to Makefile

scr2016 committed:
extended the expressions for USE_DESEL_ROWS and USE_SEL_DESEL_COLS to the second line

scr2016 committed:
dgecxx.f: added to SRC/CMakeLists.txt and SRC/lapack_64.h

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*>          matrix in the blocked step auxiliary subroutine DLAQP3RK ).
+*> \endverbatim
+*>
+*> \param[out] LIWORK
```
Corrected.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*>          The dimension of the array LIWORK. LIWORK >= N
+*>
+*>          If LIWORK = -1, then a workspace query is assumed; the routine
+*>          only calculates the optimal size of the WORK array, returns
```
Typo fixed.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*
+*     DGECX
+*
+      END 
```
Sure, fixed.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+            NBOPT = ILAENV( INB, 'DGEQRF', ' ', M, N, -1, -1 )
+            LWKOPT =  1000
+         END IF
+         WORK( 1 ) = DBLE( LWKOPT )
```
Done.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*           MSUB.
+*          
+            INFO = -6
+            WORK( 1 ) = DBLE( LWKOPT )
```
Resolved.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*         
+      END IF
+*
+      WORK( 1 ) = DBLE( LWKOPT ) 
```
Resolved.

scr2016 commented on SRC/dgecxx.f:
```diff
@@ -0,0 +1,1437 @@
+      NFREE = NSUB
+*              
+      LQUERY = ( LWORK.EQ.-1 )
+      LIQUERY = ( LIWORK.EQ.-1 )
```
Agree, and changed to   LQUERY = ( LWORK.EQ.-1 .OR. LIWORK.EQ.-1 )

Resolved.

scr2016 committed:
dgecxx.f: added quick return for (M=0 or N=0) and (M_sub=0 or N_sub=0)

If (M=0 or N=0), there is no matrix A(1:M,1:N).
If (M_sub=0 or N_sub=0), there is no matrix A_sub(1:M_sub,1:N_sub).

  modified: SRC/dgecxx.f

scr2016 committed:
dgecxx.f: added testing code for DOUBLE PRECISION.

	modified:   TESTING/LIN/CMakeLists.txt
	modified:   TESTING/LIN/Makefile
	modified:   TESTING/LIN/alaerh.f
	modified:   TESTING/LIN/alahd.f
	modified:   TESTING/LIN/dchkaa.F
	new file:   TESTING/LIN/dchkcxx.f
	new file:   TESTING/LIN/derrcxx.f
	modified:   TESTING/LIN/dlatb4.f
	modified:   TESTING/dtest.in

scr2016 commented on SRC/dgecxx.f:
```diff
@@ -0,0 +1,1437 @@
+*> \author Univ. of Colorado Denver
+*> \author NAG Ltd.
+*
+*> \ingroup gecxx
```
Added  to groups-user.dox between unitary_top @{ ... @} and geev_top @{ ... @} top-level groups 
[New added groups, lines 457-464](https://github.com/Reference-LAPACK/lapack/pull/1161/changes/1cbb16845ebd0161d3d89d4d4e9a2c956df9afc2#diff-d6fb74034dfa4963d66cf5ace96ff96f4f101c63969a288ac4f75b5fc6447edfR457-R464):
```
    @defgroup low_rank_top          Low-rank factorizations (CX, CUR, etc.)
    @{
        @defgroup cx_grp            CX factorization
        @{
            @defgroup gecxx         gecxx:          CX factorization, expert interface
        @}
    @}
```

Resolved!

scr2016 committed:
DOCS/groups-usr.dox: added new groups for CX factorization

    @defgroup low_rank_top          Low-rank factorizations (CX, CUR, etc.)
    @{
        @defgroup cx_grp            CX factorization
        @{
            @defgroup gecxx         gecxx:          CX factorization, expert interface
        @}
    @}

modified:   DOCS/groups-usr.dox

scr2016 committed:
dgecxx.f: rearranged the leading comments

modified:   SRC/dgecxx.f

scr2016 commented on SRC/dgecxx.f:
My Doxygen generation produces 2 duplicate Purpose and Parameters paragraphs with their contents, when executing in `lapack` dir: `doxygen  DOCS/Doxyfile`.
Doxygen markup in `dgecxx.f` looks correct.
What is the problem?

mgates3 replied:
Doing `doxygen DOCS/Doxyfile` or `make html` both worked for me. The dgecxx page looks normal. If you don't have `dot` installed, it won't create a call graph.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*>
+*> DGECX computes a CX factorization of a real M-by-N matrix A:
+*>
+*>   A * P(K) = C*X + A_resid, where
```
This notation was used in DGEQP3RK.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*>     from the original matrix A,
+*>
+*>   X is a K-by-N matrix that minimizes the Frobenius norm of the
+*>     residual matrix A_resid, X = pseudoinv(C) * A,
```
Either way is OK.

scr2016 commented on SRC/dgecx.f:
```diff
@@ -0,0 +1,1173 @@
+*> Column selection stage 1.
+*> =========================
+*> 
+*> The user can select N_sel columns and deselect N_desel columns
```
This is a complicated expert interface, and it has an option not to use selection-deselection, via USESD flag. A simplified interface would come later (which will not have selection-deselection).

scr2016 committed:
Optimized quick return when (M=0 or N=0), and (MSUB=0 or NSUB=0).

     Quick rfturn if possible for
     a)  M = 0 or N = 0. There is no matrix A(1:M,1:N).
     b)  MSUB = 0 or NSUB = 0. There is no matrix A_sub(1:MSUB,1:NSUB).

     modified:   SRC/dgecxx.f

mgates3 commented on SRC/lapack_64.h:
```diff
@@ -26,7 +26,7 @@
 #define CGBCON CGBCON_64
 #define CGBEQU CGBEQU_64
 #define CGBEQUB CGBEQUB_64
-#define CGBMV CGBMV_64
```
Probably don't want to remove this.

scr2016 replied:
My bad. I accidentally removed line 29.  Resolved, restored line 29. [Line 29](https://github.com/Reference-LAPACK/lapack/pull/1161/changes/83a7af2acd45edb8bcdcdfb62de33932f7acb694#diff-b35ac177f01b9d6f822eda13f423bf42c85d6850bde92e89d3ad43b85c616c30R28-R30)

scr2016 committed:
dgecxx.f: Updated comments before row deselection code.

modified:   SRC/dgecxx.f

scr2016 committed:
dgecxx.f: added 'On exit' parameter descriptions to DESEL_ROWS and SEL_DESEL_COLS, since they are modified on output.

1) added 'On exit' parameter descriptions to DESEL_ROWS and SEL_DESEL_COLS, since they are modified on output.

2) For row deselection, changed:

                   IPIV( I ) = IPIV( MSUB )
                  IPIV( MSUB ) = I
                  ITEMP = DESEL_ROWS( I )
                  DESEL_ROWS( I ) = DESEL_ROWS( MSUB )
                  DESEL_ROWS( MSUB ) = ITEMP    into:

into
                  IPIV( I ) = IPIV( MSUB )
                  IPIV( MSUB ) = I
                  DESEL_ROWS( MSUB ) = DESEL_ROWS( I )
                  DESEL_ROWS( I ) = -1

2) cleaned up comments in the code dgecxx.f

modified:   SRC/dgecxx.f

scr2016 committed:
dgecxx.f: improved algorithm how the matrix C is generated.

In case, when we return the matrrix X as well, we use the copy
of the origonal matrix A that is stored in X. So that we do not need
to save the matrix A into the array C, and do not need to select the
columns of A in the array C in place. Instead, we just copy the selected
columns of the matrix A into the matrix C.

   modified:   SRC/dgecxx.f

scr2016 committed:
SRC/lapack_64.h: restored blank line 29 to `#define CGBMV CGBMV_64`

Line 29 was emtied in error while adding

at commit
ce5b69cbf0a7ac9bc8de281d0b3c7d9a5903fb22

index 9980399df..7a7c50f83 100644
--- a/SRC/lapack_64.h
+++ b/SRC/lapack_64.h
@@ -26,7 +26,7 @@
 #define CGBCON CGBCON_64
 #define CGBEQU CGBEQU_64
 #define CGBEQUB CGBEQUB_64
-
+#define CGBMV CGBMV_64
 #define CGBRFS CGBRFS_64
 #define CGBRFSX CGBRFSX_64

modified:   SRC/lapack_64.h

scr2016 committed:
SRC/dgecxx.f: Coorrrected spelling mistakes  comments.

modified:   SRC/dgecxx.f

scr2016 committed:
SRC/dgecxx.f: Changed how the number of selected columns K is updated

after NSEL columns and NFREE columns factorization:
from update style K = NSEL,  and K = NSEL + KFREE
to increment style K = K + NSEL, K = K + KFEE

	modified:   SRC/dgecxx.f

scr2016 committed:
SRC/dgecxx.f: Simplified quick return condition.

since min( M, N) = 0 implies min( MSUB, NSUB) = 0,

changed

From:

      IF( MIN( MINMN, MIN( MSUB, NSUB )).EQ.0 ) THEN
         K = 0
         MAXC2NRMK = ZERO
         RELMAXC2NRMK = ZERO
         FNRMK = ZERO
         RETURN
      END IF
To:
     IF( MIN( MSUB, NSUB ).EQ.0 ) THEN
         K = 0
         MAXC2NRMK = ZERO
         RELMAXC2NRMK = ZERO
         FNRMK = ZERO
         RETURN
      END IF

	modified:   SRC/dgecxx.f

scr2016 committed:
SRC/dgecxx.f: Corrected the descriptions of K and TAU parameters.

	modified:   SRC/dgecxx.f

scr2016 committed:
SRC/dgecxx.f: changed TAU parameter description

	modified:   SRC/dgecxx.f

scr2016 committed:
SRC/dgecxx.f: reordered the variables in variable declarations

cosmetic change
  modified:   SRC/dgecxx.f

scr2016 committed:
added single precision file SRC/sgecxx.f

Also included sgecxx.f in to
SRC/CMakeLists.txt
SRC/Makefile

and modified SRC/lapack_64.h to reflect the addition
SRC/sgecxx.f

	modified:   SRC/CMakeLists.txt
	modified:   SRC/Makefile
	modified:   SRC/lapack_64.h
	new file:   SRC/sgecxx.f

scr2016 committed:
TESTING/LIN/dchkaa.F: fixed routine declaraiton truncation

	modified:   TESTING/LIN/dchkaa.F

scr2016 committed:
TESTING/LIN/dchkcxx.f: removed unused variable and routine declarations

also changed Definition sections in the top comments

	modified:   TESTING/LIN/dchkcxx.f

scr2016 committed:
TESTING/LIN/dchkcxx.f: removed unused function declaration fro DLANGE

	modified:   TESTING/LIN/dchkcxx.f

scr2016 committed:
TESTING/LIN/dchkcxx.f: removed more unsused routine declarations

	modified:   TESTING/LIN/dchkcxx.f

scr2016 committed:
TESTING/LIN/dchkcxx.f:  replaced constant 0D+0 with ZERO parameter

from
RESULT( 5 ) = 0D+0
to
RESULT( 5 ) = ZERO

	modified:   TESTING/LIN/dchkcxx.f

scr2016 committed:
TESTING/LIN/dchkaa.F: added DCHKCXX to routine declaration

	modified:   TESTING/LIN/dchkaa.F

scr2016 committed:
LIN/dchkcxx.f: chnaged comments for test 5.

	modified:   LIN/dchkcxx.f

scr2016 committed:
added single precision test code for SGECXX

	modified:   TESTING/LIN/CMakeLists.txt
	modified:   TESTING/LIN/Makefile
	modified:   TESTING/LIN/schkaa.F
	new file:   TESTING/LIN/schkcxx.f
	new file:   TESTING/LIN/serrcxx.f
	modified:   TESTING/LIN/slatb4.f
	modified:   TESTING/stest.in

scr2016 committed:
SRC/dgecxx.f and sgecxx.f changed formattting in description of LWORK and LIWORK

	modified:   SRC/dgecxx.f
	modified:   SRC/sgecxx.f

scr2016 committed:
SRC/dgecxx.f and sgecxx.f changed MAX to max in WORK and IWORK descript.

	modified:   SRC/dgecxx.f
	modified:   SRC/sgecxx.f

scr2016 committed:
SRC/dgecxx.f and sgecxx.f edited comments in workspace calculation

	modified:   SRC/dgecxx.f
	modified:   SRC/sgecxx.f

scr2016 committed:
SRC/zgecxx.f: added COMPLEX*16 code for CXX

	modified:   SRC/CMakeLists.txt
	modified:   SRC/Makefile
	modified:   SRC/lapack_64.h
	new file:   SRC/zgecxx.f

scr2016 committed:
TESTING/LIN/(d,s)chkcxx.f changed how workspace is calculated

scr2016 committed:
TESTING/LIN/(d,s)chkcxx.f corrected the descriptio of WORK parameter

	modified:   TESTING/LIN/dchkcxx.f
	modified:   TESTING/LIN/schkcxx.f

scr2016 committed:
TESTING/LIN/(d,s)chkcxx.f improved comments for WORK parameter
	modified:   TESTING/LIN/dchkcxx.f
        modified:   TESTING/LIN/schkcxx.f

scr2016 committed:
SRC/zgecxx.f changed the description of LWORK, and fixed  ZUNMQR transpose argument

Fixed ZUNMQR call transpose argumnet 'T' into conjugate transpose 'C'.

	modified:   SRC/zgecxx.f

scr2016 committed:
added test code for SRC/zgecxx.f

	modified:   TESTING/LIN/CMakeLists.txt
	modified:   TESTING/LIN/Makefile
	modified:   TESTING/LIN/zchkaa.F
	new file:   TESTING/LIN/zchkcxx.f
	new file:   TESTING/LIN/zerrcxx.f
	modified:   TESTING/LIN/zlatb4.f
	modified:   TESTING/ztest.in

scr2016 committed:
SRC/cgecxx.f added complex precision CX factorization routine

	modified:   SRC/CMakeLists.txt
	modified:   SRC/Makefile
	new file:   SRC/cgecxx.f
	modified:   SRC/lapack_64.h

scr2016 committed:
TESTING/LIN/zchkcxx.f: added ZERRCXX declaration

	modified:   TESTING/LIN/zchkcxx.f

scr2016 committed:
TESTING/LIN/zchkcxx.f: fixed a typo in code comments

	modified:   TESTING/LIN/zchkcxx.f

scr2016 committed:
added test code for SRC/cgecxx.f

	modified:   TESTING/LIN/CMakeLists.txt
	modified:   TESTING/LIN/Makefile
	modified:   TESTING/LIN/cchkaa.F
	new file:   TESTING/LIN/cchkcxx.f
	new file:   TESTING/LIN/cerrcxx.f
	modified:   TESTING/LIN/clatb4.f
	modified:   TESTING/ctest.in

scr2016 committed:
TESTING/LIN/schkcxx.f dchkcxx.f fixed typos in code comments

        modified:   TESTING/LIN/dchkcxx.f
        modified:   TESTING/LIN/schkcxx.f

scr2016 committed:
changed location of (Z,C)ERRCXX declaration in (z,c)errcxx.f

	modified:   TESTING/LIN/cchkcxx.f
	modified:   TESTING/LIN/zchkcxx.f

scr2016 committed:
dgecxx.f: added lapecke interface

	modified:   LAPACKE/include/lapack.h
	modified:   LAPACKE/include/lapacke.h
	modified:   LAPACKE/include/lapacke_64.h
	new file:   LAPACKE/include/lapacke_mangling.h
	modified:   LAPACKE/include/lapacke_mangling_with_flags.h.in
	modified:   LAPACKE/src/CMakeLists.txt
	modified:   LAPACKE/src/Makefile
	new file:   LAPACKE/src/lapacke_dgecxx.c
	new file:   LAPACKE/src/lapacke_dgecxx_work.c

scr2016 committed:
sgecxx.f: added lapacke interface

	modified:   LAPACKE/include/lapack.h
	modified:   LAPACKE/include/lapacke.h
	modified:   LAPACKE/include/lapacke_64.h
	modified:   LAPACKE/src/CMakeLists.txt
	modified:   LAPACKE/src/Makefile
	new file:   LAPACKE/src/lapacke_sgecxx.c
	new file:   LAPACKE/src/lapacke_sgecxx_work.c

scr2016 committed:
zgecxx.f: added lapacke interface

	modified:   LAPACKE/include/lapack.h
	modified:   LAPACKE/include/lapacke.h
	modified:   LAPACKE/include/lapacke_64.h
	modified:   LAPACKE/src/CMakeLists.txt
	modified:   LAPACKE/src/Makefile
	new file:   LAPACKE/src/lapacke_zgecxx.c
	new file:   LAPACKE/src/lapacke_zgecxx_work.c

scr2016 committed:
cgecxx.f: added lapacke interface

	modified:   LAPACKE/include/lapack.h
	modified:   LAPACKE/include/lapacke.h
	modified:   LAPACKE/include/lapacke_64.h
	modified:   LAPACKE/src/CMakeLists.txt
	modified:   LAPACKE/src/Makefile
	new file:   LAPACKE/src/lapacke_cgecxx.c
	new file:   LAPACKE/src/lapacke_cgecxx_work.c

langou committed:
Merge branch 'master' into scr2016-dgecx-draft

langou committed:
add a return to line
